Domains: Modify HeaderCart to force cart to fetch when it mounts

sirbrillig opened the pull request:
HeaderCart (added in https://github.com/Automattic/wp-calypso/pull/45339) currently does not render PopoverCart if the cart object it sees has no items. However, this can sometimes be inaccurate, since the cart object currently used by HeaderCart comes from the deprecated CartStore. When PopoverCart first mounts, it forces the CartStore to update, making sure that its data is as fresh as possible, but if PopoverCart is never rendered, that forced update will not occur, meaning that the HeaderCart could be out-of-date until the next time CartStore polls the shopping-cart endpoint.

This PR modifies HeaderCart to also force a cart fetch when it mounts. This does mean that the cart will be fetched twice if not empty, but that should be fine for the time being until PopoverCart can be refactored to work with the ShoppingCartProvider (this is in-progress).

The main purpose of this PR is to fix the `Managing Domains: (desktop) Adding a domain to an existing site Empty the cart - Empty the cart` e2e test, as discussed in https://github.com/Automattic/wp-calypso/pull/47379#issuecomment-728464578.

<img width="1086" alt="Screen Shot 2020-11-19 at 3 50 15 PM" src="https://user-images.githubusercontent.com/2036909/99722812-3ac2f300-2a7f-11eb-82a0-994330523464.png">

<img width="1108" alt="Screen Shot 2020-11-19 at 3 51 04 PM" src="https://user-images.githubusercontent.com/2036909/99722818-3dbde380-2a7f-11eb-9ddc-dfa36a52d575.png">


#### Testing instructions

- Start with an empty cart.
- Visit `/domains/add` and click to add a domain to your cart.
- You'll see the G Suite upsell page. Without confirming, click the "Domains" link in the sidebar (under "Manage").
- Verify that you see the cart icon with a "1" on it.
- Click the "Add a domain to this site" button at the top of the page.
- Verify that you see the cart icon with a "1" on it.
- Click on the cart icon and click the trash button to remove the domain from the cart.
- Verify that the cart icon disappears.

sirbrillig committed:
Modify HeaderCart to allow PopoverCart to hide itself

HeaderCart currently does not render PopoverCart if the cart object it
sees has no items. However, this can sometimes be inaccurate, since the
cart object currently used by HeaderCart comes from the deprecated
CartStore. When PopoverCart first mounts, it forces the CartStore to
update, making sure that its data is as fresh as possible, but if
PopoverCart is never rendered, that forced update will not occur,
meaning that the HeaderCart could be out-of-date until the next time
CartStore polls the shopping-cart endpoint.

Since PopoverCart already has a mechanism to hide itself via its
`visible` prop, this change modifies HeaderCart to use that prop when it
wants to hide the PopoverCart rather than not rendering it at all. This
allows the PopoverCart to fetch the latest cart even if it is invisible.

sirbrillig committed:
s/props/state

klimeryk commented on client/my-sites/checkout/cart/header-cart.jsx:
```diff
@@ -31,15 +30,18 @@ class HeaderCart extends React.Component {
-		if ( isEmpty( getAllCartItems( this.props.cart ) ) ) {
-			return null;
+		const isCartEmpty = getAllCartItems( this.props.cart ).length === 0;
+		const noop = () => {};
```
Just a quick comment that it might be better to use [lodash's `noop`](https://lodash.com/docs/4.17.15#noop). Or at least define this one as a method on this component. Otherwise, react could think it's always a new prop and re-render the cart.

sirbrillig replied:
That's an interesting point! I had removed it because I think it's a good idea to remove lodash anywhere we can, since it heavily uses its own functions and can add a lot to bundle sizes, but as this does re-create the function on each render of `HeaderCart`, it may unnecessarily re-render `PopoverCart`, but only if `PopoverCart` is memoized (passed through `connect()` or `React.memo()` or implements `shouldComponentUpdate`), which is does not. If a component is not memoized, it will always be rendered when its parent is re-rendered, even if its props do not change.

If a component's output does not change, React will not update the DOM, so the main reason that I know of to avoid extra renders is expensive side effects, but I don't see any such side effects in `PopoverCart`, and this change shouldn't cause more renders in any case.

That said, it really seems to me that `closeSectionNavMobilePanel` should be optional on `PopoverCart` (the PropTypes on that component list it as optional, but it's actually required). I may push a simple change as part of this PR so that it doesn't need a noop at all.

klimeryk replied:
TIL! Thanks for the in-depth explanation 🙇

sirbrillig replied:
For completeness: there's one other case I thought of that can prevent rendering if the props don't change: if the component extends `React.PureComponent`.

sirbrillig committed:
Stop passing noop closeSectionNavMobilePanel to PopoverCart

sirbrillig committed:
Only call closeSectionNavMobilePanel in PopoverCart if present

The prop is listed as optional, so we should make sure that missing it
does not cause a fatal.

michaeldcain reviewed (changes requested):
Code-wise, this looks good to me, and the failing tests seem to be unrelated.

> while retaining the same behavior

This doesn't seem to be quite the case. With these changes, an empty cart will still render the cart icon, even when the PopoverCart isn't pinned. I believe you can fix this by additionally [checking if the `visible` prop is true here](https://github.com/Automattic/wp-calypso/blob/update%2Fheader-cart-to-allow-popover-cart-to-hide-itself/client/my-sites/checkout/cart/popover-cart.jsx#L127).

sirbrillig commented:
> This doesn't seem to be quite the case. With these changes, an empty cart will still render the cart icon, even when the PopoverCart isn't pinned. I believe you can fix this by additionally checking if the visible prop is true here.

🤔  I may be misunderstanding... if I'm reading correctly you're saying that if the `pinned` prop is false (which it always is in this case), and the cart is empty (causing the `visible` to be false), `PopoverCart` will still render?

sirbrillig commented:
Ah, I was wrong! The `visible` prop means "is the popup contents visible", not "is the icon button visible". I'll fix this up. Thanks for noticing!

sirbrillig committed:
Force cart to reload when HeaderCart mounts

sirbrillig commented:
I refactored this PR to instead force the cart to update in HeaderCart as well as PopoverCart.

michaeldcain reviewed (approved):
:shipit:
